misc/multistream-select/src/dialer_select.rs

```diff
@@ -268,6 +450,182 @@ where
+                    if let Err(err) = Pin::new(&mut io).start_send(msg) {
+                        return Poll::Ready(Err(From::from(err)));
+                    }
+
+                    self.state = SimOpenState::FlushNonce { io, local_nonce };
+                }
+                SimOpenState::FlushNonce {
+                    mut io,
+                    local_nonce,
+                } => match Pin::new(&mut io).poll_flush(cx)? {
+                    Poll::Ready(()) => self.state = SimOpenState::ReadNonce { io, local_nonce },
+                    Poll::Pending => {
+                        self.state = SimOpenState::FlushNonce { io, local_nonce };
+                        return Poll::Pending;
+                    }
+                },
+                SimOpenState::ReadNonce {
+                    mut io,
+                    local_nonce,
+                } => {
+                    let msg = match Pin::new(&mut io).poll_next(cx)? {
+                        Poll::Ready(Some(msg)) => msg,
+                        Poll::Pending => {
+                            self.state = SimOpenState::ReadNonce { io, local_nonce };
+                            return Poll::Pending;
+                        }
+                        // Treat EOF error as [`NegotiationError::Failed`], not as
+                        // [`NegotiationError::ProtocolError`], allowing dropping or closing an I/O
+                        // stream as a permissible way to "gracefully" fail a negotiation.
+                        Poll::Ready(None) => return Poll::Ready(Err(NegotiationError::Failed)),
+                    };
+
+                    match msg {
+                        // As an optimization, the simultaneous open
+                        // multistream-select variant sends both the
+                        // simultaneous open ID (`/libp2p/simultaneous-connect`)
+                        // and a protocol before flushing. In the case where the
+                        // remote acts as a listener already, it can accept or
+                        // decline the attached protocol within the same
+                        // round-trip.
+                        //
+                        // In this particular situation, the remote acts as a
+                        // dialer and uses the simultaneous open variant. Given
+                        // that nonces need to be exchanged first, the attached
+                        // protocol by the remote needs to be ignored.
+                        Message::Protocol(_) => {
+                            self.state = SimOpenState::ReadNonce { io, local_nonce };
+                        }
+                        Message::Select(remote_nonce) => {
+                            match local_nonce.cmp(&remote_nonce) {
+                                Ordering::Equal => {
+                                    // Start over.
+                                    self.state = SimOpenState::SendNonce { io };
```

Comment:
The [spec](
https://github.com/libp2p/specs/blob/master/connections/simopen.md#the-protocol) says we should fail in this case!

> In the unlikely case where both peers selected the same integer, connection establishment fails.